### PR TITLE
DI-98 Log details of unmatched pharmacies into splunk

### PR DESCRIPTION
**Pull request by rohith1122**
## Link to JIRA Ticket

- https://nhsd-jira.digital.nhs.uk/browse/DI-98

## Description
Log unmatched pharmacy services that have come through into DI from nhs.uk in order to compile an ad hoc list for reporting 

### Noteworthy Changes

- Added separate function to log  unmatched  pharmacies
- Added unit tests 
- New feature (non-breaking change which adds functionality)

## Testing

Please tick the testing that has been completed

- [ ] Unit tests
- [ ] Component tests

## Developer Checklist

- [x] I have performed a self-review of my own code
- [x] I have run the [code formatting checks](../README.md#code-quality)
- [x] I have run the [code quality checks](../README.md#code-quality)
- [ ] New code meets [standards](https://nhsd-confluence.digital.nhs.uk/display/DI/DI+Ways+of+Working) agreed by the team
- [x] Unit test code coverage is at or above 80%
- [x] New and existing unit tests pass locally with my changes
- [ ] Tests have added that prove my fix is effective or that my feature works (Component tests)
- [ ] I have made corresponding changes to the documentation
- [ ] I have cleaned down my environment (if created)

## Code Reviewer Checklist

- [ ] I have run the unit tests and they run correctly
- [ ] I can confirm the changes have been tested or approved by a tester
- [ ] I can confirm no remaining infrastructure is left over from this branch

### CODE REVIEW

File: application/event_processor/event_processor.py [hunk 1]
```diff
@@ -190,6 +191,30 @@ def update_changes_with_opening_times(changes: dict, dos_service: DoSService, nh
         changes[OPENING_DAYS_KEY] = nhs_std_open_dates.export_cr_format()
 
 
+def log_unmatched_nhsuk_pharmacies(nhs_entity: NHSEntity) -> None:
+    """Log unmatched NHS Pharmacies
+    Args:
+        nhs_entity (NHSEntity): NHS entity to log
+    """
+    logger.info(
```
Comment (ed-sparkes) [inline application/event_processor/event_processor.py:199]: Maybe a logger.warning is more appropriate?

File: application/event_processor/event_processor.py [hunk 2]
```diff
@@ -190,6 +191,30 @@ def update_changes_with_opening_times(changes: dict, dos_service: DoSService, nh
...
+        "No matching DOS services found in DOS for a given NHS pharmacy service",
+        extra={
+            "log_id": UN_MATCHED_PHARMACY_LOG_ID,
```
Comment (ed-sparkes) [inline application/event_processor/event_processor.py:202]: @JackPlowman  used report_key in his PR can we keep the key field name consistent

File: application/event_processor/event_processor.py [hunk 3]
```diff
@@ -29,6 +29,7 @@
 tracer = Tracer()
 lambda_client = boto3.client("lambda")
 expected_env_vars = ("DB_SERVER", "DB_PORT", "DB_NAME", "DB_SCHEMA", "DB_USER_NAME", "EVENT_SENDER_LAMBDA_NAME")
+UN_MATCHED_PHARMACY_LOG_ID = "UN_MATCHED_PHARMACY"
```
Comment (JackPlowman) [inline application/event_processor/event_processor.py:32]: Can we move all the reporting/logging functionality into a new/another file called something like `reporting.py`. We will eventually have many reporting/logging functions so it isn't feasible to keep them all in this `event_processor.py` file. I've done the same in DI-99 and am happy to rebase off whatever you do if this PR does in first.

File: application/event_processor/event_processor.py [hunk 4]
```diff
@@ -220,6 +245,7 @@ def lambda_handler(event: Dict[str, Any], context: LambdaContext) -> None:
     matching_services = event_processor.get_matching_services()
 
     if len(matching_services) == 0:
+        log_unmatched_nhsuk_pharmacies(nhs_entity)
         logger.error(f"No matching DOS services found that fit all " f"criteria for ODSCode '{nhs_entity.ODSCode}'")
```
Comment (JackPlowman) [inline application/event_processor/event_processor.py:249]: Is there any benefit to this line as you have a replacement in the function below - this line `No matching DOS services found in DOS for a given NHS pharmacy service`
``` 
logger.info(
        "No matching DOS services found in DOS for a given NHS pharmacy service",
        extra={
            "log_id": UN_MATCHED_PHARMACY_LOG_ID,
            "nhsuk_odscode": nhs_entity.ODSCode,
            "nhsuk_org_name": nhs_entity.OrganisationName,
            "nhsuk_service_type": nhs_entity.ServiceType,
            "nhsuk_service_status": nhs_entity.OrganisationStatus,
            "nhsuk_service_sub_sector": nhs_entity.OrganisationSubType,
            "nhsuk_address1": nhs_entity.Address1 if hasattr(nhs_entity, "Address1") else "",
            "nhsuk_address2": nhs_entity.Address2 if hasattr(nhs_entity, "Address2") else "",
            "nhsuk_address3": nhs_entity.Address3 if hasattr(nhs_entity, "Address3") else "",
            "nhsuk_address4": nhs_entity.Address4 if hasattr(nhs_entity, "Address4") else "",
            "nhsuk_address5": nhs_entity.Address5 if hasattr(nhs_entity, "Address5") else "",
            "nhsuk_postcode": nhs_entity.Postcode,
        },
    )
```

File: application/event_processor/tests/test_event_processor.py [hunk 5]
```diff
@@ -372,3 +381,41 @@ def test_update_changes_address_to_change_request_if_not_equal_not_equal():
     actual_changes = update_changes_with_address(changes, "address", dos_address, nhs_uk_entity)
     # Assert
     assert expected_changes == actual_changes, f"Should return {changes} dict, actually: {actual_changes}"
+
+
+@patch.object(Logger, "info")
```
Comment (JackPlowman) [inline application/event_processor/tests/test_event_processor.py:386]: Great unit test, if you shift the function to another file can a test file be created with this unit test in as well

**Review by JackPlowman — CHANGES_REQUESTED** [review]
Couple of small changes, feel free to disagree with them

File: application/event_processor/event_processor.py (hunk 3 above)
Comment (rohith1122) [reply application/event_processor/event_processor.py:32]: I was thinking the same. completely agree with you. I will wait for you to merge first and then rebase, which is easy for me.

File: application/event_processor/event_processor.py (hunk 4 above)
Comment (rohith1122) [reply application/event_processor/event_processor.py:249]: Yes, kind of duplicating the same message.  I will remove "No matching DOS services  ...."

File: application/event_processor/tests/test_event_processor.py (hunk 5 above)
Comment (rohith1122) [reply application/event_processor/tests/test_event_processor.py:386]: Sure . agree

**Review by ed-sparkes — CHANGES_REQUESTED** [review]

File: application/event_processor/event_processor.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rohith1122) [reply application/event_processor/event_processor.py:249]: Sure

File: application/event_processor/event_processor.py (hunk 2 above)
Comment (rohith1122) [reply application/event_processor/event_processor.py:202]: Just had a look at @JackPlowman  PR and  will follow the same pattern

File: application/event_processor/event_processor.py (hunk 1 above)
Comment (rohith1122) [reply application/event_processor/event_processor.py:199]: Sure

**Review by ed-sparkes — DISMISSED** [review]

**Review by ed-sparkes — DISMISSED** [review]

**Review by JackPlowman — DISMISSED** [review]
LGTM 👍

**Review by JackPlowman — APPROVED** [review]
LGTM 👍

**Review by ed-sparkes — APPROVED** [review]

**Pull request merged**